### ISSUE TITLE
`ReadsSparkSource`  cannot (seem to) read cram file on hdfs or gcs. 

### ISSUE THREAD

**Issue by SHuang-Broad** [broadinstitute/gatk#2382]
While trying to run a tool with a CRAM input located on a hdfs attached to a dataproc cluster, we are hit by an error message described below:

```
***********************************************************************

A USER ERROR has occurred: Failed to read bam header from hdfs://svdev-caller-m:8020/data/smallCram.cram
 Caused by:Pathname /tmp/da63aa3c-e3bc-4893-9f40-42921719a343/hdfs:/svdev-caller-m:8020/reference/Homo_sapiens_assembly38.fasta from /tmp/da63aa3c-e3bc-4893-9f40-42921719a343/hdfs:/svdev-caller-m:8020/reference/Homo_sapiens_assembly38.fasta is not a valid DFS filename.

***********************************************************************
org.broadinstitute.hellbender.exceptions.UserException: A USER ERROR has occurred: Failed to read bam header from hdfs://svdev-caller-m:8020/data/smallCram.cram
 Caused by:Pathname /tmp/da63aa3c-e3bc-4893-9f40-42921719a343/hdfs:/svdev-caller-m:8020/reference/Homo_sapiens_assembly38.fasta from /tmp/da63aa3c-e3bc-4893-9f40-42921719a343/hdfs:/svdev-caller-m:8020/reference/Homo_sapiens_assembly38.fasta is not a valid DFS filename.
	at org.broadinstitute.hellbender.engine.spark.datasources.ReadsSparkSource.getHeader(ReadsSparkSource.java:206)
	at org.broadinstitute.hellbender.engine.spark.GATKSparkTool.initializeReads(GATKSparkTool.java:381)
	at org.broadinstitute.hellbender.engine.spark.GATKSparkTool.initializeToolInputs(GATKSparkTool.java:361)
	at org.broadinstitute.hellbender.engine.spark.GATKSparkTool.runPipeline(GATKSparkTool.java:351)
	at org.broadinstitute.hellbender.engine.spark.SparkCommandLineProgram.doWork(SparkCommandLineProgram.java:38)
	at org.broadinstitute.hellbender.cmdline.CommandLineProgram.runTool(CommandLineProgram.java:112)
	at org.broadinstitute.hellbender.cmdline.CommandLineProgram.instanceMainPostParseArgs(CommandLineProgram.java:170)
	at org.broadinstitute.hellbender.cmdline.CommandLineProgram.instanceMain(CommandLineProgram.java:189)
	at org.broadinstitute.hellbender.Main.instanceMain(Main.java:96)
	at org.broadinstitute.hellbender.Main.instanceMain(Main.java:103)
	at org.broadinstitute.hellbender.Main.mainEntry(Main.java:116)
	at org.broadinstitute.hellbender.Main.main(Main.java:158)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.apache.spark.deploy.SparkSubmit$.org$apache$spark$deploy$SparkSubmit$$runMain(SparkSubmit.scala:738)
	at org.apache.spark.deploy.SparkSubmit$.doRunMain$1(SparkSubmit.scala:187)
	at org.apache.spark.deploy.SparkSubmit$.submit(SparkSubmit.scala:212)
	at org.apache.spark.deploy.SparkSubmit$.main(SparkSubmit.scala:126)
	at org.apache.spark.deploy.SparkSubmit.main(SparkSubmit.scala)
Caused by: java.lang.IllegalArgumentException: Pathname /tmp/da63aa3c-e3bc-4893-9f40-42921719a343/hdfs:/svdev-caller-m:8020/reference/Homo_sapiens_assembly38.fasta from /tmp/da63aa3c-e3bc-4893-9f40-42921719a343/hdfs:/svdev-caller-m:8020/reference/Homo_sapiens_assembly38.fasta is not a valid DFS filename.
	at org.apache.hadoop.hdfs.DistributedFileSystem.getPathName(DistributedFileSystem.java:213)
	at org.apache.hadoop.hdfs.DistributedFileSystem$27.doCall(DistributedFileSystem.java:1436)
	at org.apache.hadoop.hdfs.DistributedFileSystem$27.doCall(DistributedFileSystem.java:1433)
	at org.apache.hadoop.fs.FileSystemLinkResolver.resolve(FileSystemLinkResolver.java:81)
	at org.apache.hadoop.hdfs.DistributedFileSystem.getFileStatus(DistributedFileSystem.java:1448)
	at org.apache.hadoop.fs.FileSystem.exists(FileSystem.java:1436)
	at org.broadinstitute.hellbender.utils.spark.SparkUtils.pathExists(SparkUtils.java:100)
	at org.broadinstitute.hellbender.engine.spark.datasources.ReadsSparkSource.setHadoopBAMConfigurationProperties(ReadsSparkSource.java:241)
	at org.broadinstitute.hellbender.engine.spark.datasources.ReadsSparkSource.getHeader(ReadsSparkSource.java:203)
	... 20 more
ERROR: (gcloud.dataproc.jobs.submit.spark) Job [da63aa3c-e3bc-4893-9f40-42921719a343] entered state [ERROR] while waiting for [DONE].
```

to reproduce this error, 

```bash
cd /Users/shuang/GATK/gatk

CLUSTER_NAME="svdev-caller"
MASTER_NODE="hdfs://svdev-caller-m:8020"
PROJECT_DIR="user/shuang/NA12878_PCR-_30X"

./gatk-launch FindBreakpointEvidenceSpark \
    -R "$MASTER_NODE"/reference/Homo_sapiens_assembly38.fasta \
    -I "$MASTER_NODE"/data/smallCram.cram \
    -O "$MASTER_NODE"/"$PROJECT_DIR"/fastq \
    --exclusionIntervals gs://sv-data-dsde-dev/reference/GRCh38.kill.intervals \
    --kmersToIgnore gs://sv-data-dsde-dev/reference/Homo_sapiens_assembly38.dups \
    --kmerIntervals "$MASTER_NODE"/"$PROJECT_DIR"/kmerIntervals \
    --breakpointEvidenceDir "$MASTER_NODE"/"$PROJECT_DIR"/evidence \
    --breakpointIntervals "$MASTER_NODE"/"$PROJECT_DIR"/intervals \
    --qnameIntervalsMapped "$MASTER_NODE"/"$PROJECT_DIR"/qnameIntervalsMapped \
    --qnameIntervalsForAssembly "$MASTER_NODE"/"$PROJECT_DIR"/qnameIntervalsForAssembly \
    --maxFASTQSize 10000000 \
    -- \
    --sparkRunner GCS \
    --cluster svdev-caller
```

========================

On the other hand, we see a similar error if the input is changed to the same file but stored in a google bucket (although the cited cause is different):

```
***********************************************************************

A USER ERROR has occurred: Failed to read bam header from gs://sv-data-dsde-dev/test_data/smallCram.cram
 Caused by:null

***********************************************************************
org.broadinstitute.hellbender.exceptions.UserException: A USER ERROR has occurred: Failed to read bam header from gs://sv-data-dsde-dev/test_data/smallCram.cram
 Caused by:null
	at org.broadinstitute.hellbender.engine.spark.datasources.ReadsSparkSource.getHeader(ReadsSparkSource.java:182)
	at org.broadinstitute.hellbender.engine.spark.GATKSparkTool.initializeReads(GATKSparkTool.java:381)
	at org.broadinstitute.hellbender.engine.spark.GATKSparkTool.initializeToolInputs(GATKSparkTool.java:361)
	at org.broadinstitute.hellbender.engine.spark.GATKSparkTool.runPipeline(GATKSparkTool.java:351)
	at org.broadinstitute.hellbender.engine.spark.SparkCommandLineProgram.doWork(SparkCommandLineProgram.java:38)
	at org.broadinstitute.hellbender.cmdline.CommandLineProgram.runTool(CommandLineProgram.java:112)
	at org.broadinstitute.hellbender.cmdline.CommandLineProgram.instanceMainPostParseArgs(CommandLineProgram.java:170)
	at org.broadinstitute.hellbender.cmdline.CommandLineProgram.instanceMain(CommandLineProgram.java:189)
	at org.broadinstitute.hellbender.Main.instanceMain(Main.java:96)
	at org.broadinstitute.hellbender.Main.instanceMain(Main.java:103)
	at org.broadinstitute.hellbender.Main.mainEntry(Main.java:116)
	at org.broadinstitute.hellbender.Main.main(Main.java:158)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.apache.spark.deploy.SparkSubmit$.org$apache$spark$deploy$SparkSubmit$$runMain(SparkSubmit.scala:738)
	at org.apache.spark.deploy.SparkSubmit$.doRunMain$1(SparkSubmit.scala:187)
	at org.apache.spark.deploy.SparkSubmit$.submit(SparkSubmit.scala:212)
	at org.apache.spark.deploy.SparkSubmit$.main(SparkSubmit.scala:126)
	at org.apache.spark.deploy.SparkSubmit.main(SparkSubmit.scala)
Caused by: java.lang.NullPointerException
	at java.io.ByteArrayInputStream.<init>(ByteArrayInputStream.java:106)
	at org.broadinstitute.hellbender.engine.AuthHolder.getOfflineAuth(AuthHolder.java:79)
	at org.broadinstitute.hellbender.engine.AuthHolder.makeStorageClient(AuthHolder.java:94)
	at org.broadinstitute.hellbender.engine.spark.datasources.ReadsSparkSource.getHeader(ReadsSparkSource.java:177)
	... 20 more
ERROR: (gcloud.dataproc.jobs.submit.spark) Job [bd000687-f538-4201-b888-668612d46bad] entered state [ERROR] while waiting for [DONE].
```

=========================

On a third note, if the reference is also provided with a GCS path, we see this:

```
***********************************************************************

A USER ERROR has occurred: The specified fasta file (gs://sv-data-dsde-dev/reference/Homo_sapiens_assembly38.fasta) does not exist.

***********************************************************************
org.broadinstitute.hellbender.exceptions.UserException$MissingReference: A USER ERROR has occurred: The specified fasta file (gs://sv-data-dsde-dev/reference/Homo_sapiens_assembly38.fasta) does not exist.
	at org.broadinstitute.hellbender.engine.datasources.ReferenceFileSource.<init>(ReferenceFileSource.java:31)
	at org.broadinstitute.hellbender.engine.datasources.ReferenceMultiSource.<init>(ReferenceMultiSource.java:49)
	at org.broadinstitute.hellbender.engine.spark.GATKSparkTool.initializeReference(GATKSparkTool.java:394)
	at org.broadinstitute.hellbender.engine.spark.GATKSparkTool.initializeToolInputs(GATKSparkTool.java:360)
	at org.broadinstitute.hellbender.engine.spark.GATKSparkTool.runPipeline(GATKSparkTool.java:351)
	at org.broadinstitute.hellbender.engine.spark.SparkCommandLineProgram.doWork(SparkCommandLineProgram.java:38)
	at org.broadinstitute.hellbender.cmdline.CommandLineProgram.runTool(CommandLineProgram.java:112)
	at org.broadinstitute.hellbender.cmdline.CommandLineProgram.instanceMainPostParseArgs(CommandLineProgram.java:170)
	at org.broadinstitute.hellbender.cmdline.CommandLineProgram.instanceMain(CommandLineProgram.java:189)
	at org.broadinstitute.hellbender.Main.instanceMain(Main.java:96)
	at org.broadinstitute.hellbender.Main.instanceMain(Main.java:103)
	at org.broadinstitute.hellbender.Main.mainEntry(Main.java:116)
	at org.broadinstitute.hellbender.Main.main(Main.java:158)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.apache.spark.deploy.SparkSubmit$.org$apache$spark$deploy$SparkSubmit$$runMain(SparkSubmit.scala:738)
	at org.apache.spark.deploy.SparkSubmit$.doRunMain$1(SparkSubmit.scala:187)
	at org.apache.spark.deploy.SparkSubmit$.submit(SparkSubmit.scala:212)
	at org.apache.spark.deploy.SparkSubmit$.main(SparkSubmit.scala:126)
	at org.apache.spark.deploy.SparkSubmit.main(SparkSubmit.scala)
ERROR: (gcloud.dataproc.jobs.submit.spark) Job [1c0c33a8-53ac-4407-a452-bb8622fd3060] entered state [ERROR] while waiting for [DONE].
```

**Comment by SHuang-Broad** [comment]
poking the engine team @lbergelson @cmnbroad @droazen

**Comment by cmnbroad** [comment]
Looks like the HDFS URI is getting prefixed. Will take a look.

**Comment by cmnbroad** [comment]
@SHuang-Broad can you let me know which build this was. Thx.

**Comment by SHuang-Broad** [comment]
@cmnbroad , it's master.

**Comment by cmnbroad** [comment]
This is happening because there are code path(s) in GATKSparkTool that retrieve the reference via getReferenceFile() (and call getAbsolutePath()). We could fix this specific case, but the upcoming URI class should provide a more general solution.

We should also add a test case that uses the hdfs mini cluster to test an actual tool (we have them for ReadsDataSource, but that bypasses the failing code path).

**Comment by SHuang-Broad** [comment]
@cmnbroad thanks for the update! This is not a urgent request, but just something nice to have as we might be dealing with more cram files, which will not be converted bams if we could read crams.

**Comment by psendil** [comment]
Seeing same issue, I'm running this from master node 

-->  ./gatk-launch PrintReadsSpark -I hdfs://gatk/input.bam -O hdfs://gatk/output.bam

getting following error:
***********************************************************************

A USER ERROR has occurred: Failed to read bam header from hdfs://gatk/input.bam
 Caused by:java.net.UnknownHostException: gatk

***********************************************************************

**Comment by cmnbroad** [comment]
Fixed by several PRs, with a test in #6618.
